model/src/main/kotlin/Scope.kt

```diff
@@ -90,4 +91,42 @@ data class Scope(
 
         return getTreeDepthRec(dependencies)
     }
+
+    /**
+     * Return the shortest path to each dependency in this scope. The path to a dependency is defined by the nodes of
+     * the dependency tree that need to be passed to get to the dependency. For direct dependencies the shortest path is
+     * empty.
+     */
+    @JsonIgnore
+    fun getShortestPaths(): Map<Identifier, List<Identifier>> {
+        data class QueueItem(
+            val pkgRef: PackageReference,
+            val parents: List<Identifier>
+        )
+
+        val remainingIds = collectDependencies().toMutableSet()
```

Comment:
Instead of keeping `remaining` ids I think one should keep have a map:
`visitedIds Map<Identifier, Identifier> mapping any already `deuqueued` node to its parent (or its identity) to improve on memory consumption, speed and consistency - see above comment.

Reply:
I did it like this to be able to verify that all IDs have been assigned a path.